libp2pctl/server.go

```diff
@@ -0,0 +1,352 @@
+
+// EndpointKey is a libp2p connection endpoint, usable as a map key.
+type EndpointKey struct {
+	id   peer.ID
+	addr string
+}
+
+// ID returns the endpoint ID.
+func (k EndpointKey) ID() (EndpointID, error) {
+	addr, err := ma.NewMultiaddrBytes([]byte(k.addr))
+	if err != nil {
+		return EndpointID{}, errors.Wrap(err, "cannot convert multiaddr")
+	}
+	return EndpointID{k.id, addr}, nil
+}
+
+// ConnID is the identifier of a connection.
+type ConnID struct {
+	Local, Remote EndpointID
+}
+
+// Key returns a struct usable as a map key.
+func (n ConnID) Key() ConnKey {
+	return ConnKey{n.Local.Key(), n.Remote.Key()}
+}
+
+// ConnKey is a connection identifier, usable as a map key.
+type ConnKey struct {
+	local, remote EndpointKey
+}
+
+// ID returns the connection ID.
+func (k ConnKey) ID() (ConnID, error) {
+	local, err := k.local.ID()
+	if err != nil {
+		return ConnID{}, errors.Wrap(err, "cannot convert local key into ID")
+	}
+	remote, err := k.remote.ID()
+	if err != nil {
+		return ConnID{}, errors.Wrap(err, "cannot convert remote key into ID")
+	}
+	return ConnID{local, remote}, nil
+}
+
+// ConnIDFromConn returns the name of the given connection.
+func ConnIDFromConn(conn network.Conn) ConnID {
+	return ConnID{
+		Local:  EndpointID{conn.LocalPeer(), conn.LocalMultiaddr()},
+		Remote: EndpointID{conn.RemotePeer(), conn.RemoteMultiaddr()},
+	}
+}
+
+// Instance is a libp2pctl instance.
+type Instance struct {
+	host     host.Host
+	rtr      *mux.Router
+	conns    map[ConnKey]network.Conn
+	connsMtx sync.RWMutex
+	notifiee *notifiee
+}
+
+// New returns a libp2pctl instance associated with the given libp2p host.
+func New(host host.Host) *Instance {
+	rtr := mux.NewRouter()
+	rtrConn := rtr.Path("/conn/{local}/{remote}").Subrouter()
+	inst := &Instance{
+		host:  host,
+		rtr:   rtr,
+		conns: map[ConnKey]network.Conn{},
+	}
+	for _, conn := range host.Network().Conns() {
+		inst.addConn(conn)
+	}
+	inst.notifiee = &notifiee{inst}
+	host.Network().Notify(inst.notifiee)
+
+	rtr.Methods("GET").Path("/addrs").HandlerFunc(inst.getAddrs)
+	rtr.Methods("GET").Path("/conns").HandlerFunc(inst.getConns)
+	rtr.Methods("POST").Path("/conns").HandlerFunc(inst.postConns)
+	rtrConn.Methods("GET").Path("").HandlerFunc(inst.getConn).Name("conn")
+	rtrConn.Methods("DELETE").Path("").HandlerFunc(inst.deleteConn)
+	rtrConn.Methods("GET").Path("/stat").HandlerFunc(inst.getConnStat)
+
+	return inst
+}
+
+func (inst *Instance) addConn(conn network.Conn) {
+	inst.connsMtx.Lock()
+	defer inst.connsMtx.Unlock()
+	inst.conns[ConnIDFromConn(conn).Key()] = conn
+}
+
+func (inst *Instance) delConn(conn network.Conn) {
+	inst.connsMtx.Lock()
+	defer inst.connsMtx.Unlock()
+	delete(inst.conns, ConnIDFromConn(conn).Key())
+}
+
+// Handler returns an HTTP request handler of the libp2pctl instance.
+func (inst *Instance) Handler() http.Handler {
+	return inst.rtr
+}
+
+func httpWriteJSON(w http.ResponseWriter, data interface{}) error {
+	bytes, err := json.Marshal(data)
```

Comment:
can make shorter with: json.NewEncoder(w).Encode(data)

Reply:
Wanted to limit truncated output case (the one-liner cannot emit HTTP 500 upon encode error).